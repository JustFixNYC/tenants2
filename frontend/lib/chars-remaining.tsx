import React from 'react';
import { SimpleProgressiveEnhancement } from "./progressive-enhancement";

/**
 * Once the user has this percentage of their maximum limit left,
 * we will be more noticeable.
 */
const DANGER_ALERT_PCT = 0.10;

export type CharsRemainingProps = {
  max: number,
  current: number
};

export function CharsRemaining({ max, current }: CharsRemainingProps): JSX.Element {
  const remaining = max - current;
  const className = remaining < (max * DANGER_ALERT_PCT) ? 'has-text-danger' : '';
  const text = `${remaining} character${remaining === 1 ? '' : 's'} remaining.`;

  return (
    <SimpleProgressiveEnhancement>
      <p className={className}>{text}</p>
    </SimpleProgressiveEnhancement>
  );
}
