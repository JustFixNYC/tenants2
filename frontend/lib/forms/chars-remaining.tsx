import React from "react";
import { SimpleProgressiveEnhancement } from "../ui/progressive-enhancement";
import {
  TextualFormFieldProps,
  TextareaFormField,
  TextualFormField,
} from "./form-fields";
import { Trans, Plural } from "@lingui/macro";

/**
 * Once the user has this percentage of their maximum limit left,
 * we will be more noticeable.
 */
const DANGER_ALERT_PCT = 0.1;

/**
 * If the user has at most this many characters left, we will be
 * more noticeable.
 */
const DANGER_ALERT_MIN_CHARS = 10;

export type CharsRemainingProps = {
  max: number;
  current: number;
  useSpan?: boolean;
};

export function CharsRemaining({
  max,
  current,
  useSpan,
}: CharsRemainingProps): JSX.Element {
  const remaining = max - current;
  const isNoticeable =
    remaining < max * DANGER_ALERT_PCT || remaining <= DANGER_ALERT_MIN_CHARS;
  const el = (
    <Trans
      render={useSpan ? "span" : "p"}
      className={isNoticeable ? "has-text-danger" : undefined}
    >
      <Plural
        value={remaining}
        one="1 character remaining"
        other="# characters remaining"
      />
    </Trans>
  );

  return <SimpleProgressiveEnhancement>{el}</SimpleProgressiveEnhancement>;
}

export function TextareaWithCharsRemaining(
  props: TextualFormFieldProps & {
    maxLength: number;
  }
) {
  return (
    <>
      <TextareaFormField {...props} />
      <CharsRemaining max={props.maxLength} current={props.value.length} />
    </>
  );
}

export function TextualFieldWithCharsRemaining(
  props: TextualFormFieldProps & {
    maxLength: number;
  }
) {
  return (
    <>
      <TextualFormField
        {...props}
        help={
          <CharsRemaining
            useSpan
            max={props.maxLength}
            current={props.value.length}
          />
        }
      />
    </>
  );
}
