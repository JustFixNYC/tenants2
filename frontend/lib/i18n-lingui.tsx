import React, { useMemo } from "react";
import { Catalog } from "@lingui/core";
import loadable, { LoadableLibrary } from "@loadable/component";
import { I18nProvider } from "@lingui/react";
import i18n, { SupportedLocale } from "./i18n";
import { setupI18n as linguiSetupI18n, Catalogs } from "@lingui/core";

/**
 * We use code splitting to make sure that we only load message
 * catalogs needed for our currently selected locale.
 *
 * This defines the type of component whose children prop is a
 * callable that receives a Lingui message catalog as its only
 * argument.
 */
export type LoadableCatalog = LoadableLibrary<Catalog>;

/**
 * Maps supported locales to components that load a Lingui message
 * catalog for them.
 */
export type LinguiCatalogMap = {
  [k in SupportedLocale]: LoadableCatalog;
};

/**
 * The "base" catalog, which contains the most common strings.
 */
const BaseCatalogMap: LinguiCatalogMap = {
  en: loadable.lib(() => import("../../locales/en/base.chunk") as any),
  es: loadable.lib(() => import("../../locales/es/base.chunk") as any),
};

const SetupI18n: React.FC<
  LinguiI18nProps & {
    locale: SupportedLocale;
    catalog: Catalog;
  }
> = (props) => {
  const { locale, catalog } = props;

  // This useMemo() call might be overkill. -AV
  const ourLinguiI18n = useMemo(() => {
    mergeIntoLinguiCatalog(locale, catalog);
    return li18n;
  }, [locale, catalog]);

  return (
    <I18nProvider language={locale} i18n={ourLinguiI18n}>
      {props.children}
    </I18nProvider>
  );
};

export type LinguiI18nProps = {
  /** Children to render once localization data is loaded. */
  children: React.ReactNode;
};

/**
 * Loads the Lingui base message catalog for the currently selected
 * locale, as dictated by our global i18n module. Children
 * will then be rendered with the catalog loaded and ready
 * to translate.
 *
 * While a loading message will appear while the catalog is being loaded,
 * because we do server-side rendering and pre-load JS bundles in the
 * server-rendered HTML output, the user won't see the message most
 * (possibly all) of the time.
 *
 * Note that this component is currently a singleton; more than one
 * instance of it should never exist in a component tree at once.
 */
export const LinguiI18n: React.FC<LinguiI18nProps> = (props) => {
  const locale = i18n.locale;
  const Catalog = BaseCatalogMap[locale];

  return (
    <Catalog fallback={<p>Loading locale data...</p>}>
      {(catalog) => <SetupI18n {...props} locale={locale} catalog={catalog} />}
    </Catalog>
  );
};

/**
 * Creates a component that will load an auxiliary message catalog for
 * Lingui in the currently selected locale, as dictated by our global i18n module.
 */
export function createLinguiCatalogLoader(
  catalogMap: LinguiCatalogMap
): React.FC<LinguiI18nProps> {
  return (props) => {
    const locale = i18n.locale;
    const Catalog = catalogMap[locale];

    return (
      <Catalog fallback={<p>Loading locale data...</p>}>
        {(catalog) => {
          mergeIntoLinguiCatalog(locale, catalog);
          return props.children;
        }}
      </Catalog>
    );
  };
}

/**
 * A global instance of Lingui's I18n object, which can be used to perform
 * localization outside of React JSX. Note that this object is populated
 * by the <LinguiI18n> component, however, so it should only really be
 * used by components that exist below it in the hierarchy.
 */
export const li18n = linguiSetupI18n();

/**
 * Internal global we maintain to keep track of all the messages we
 * can translate across our various catalog chunks.
 */
const catalogs: Catalogs = {};

/**
 * Merge the given catalog for the given locale into our global
 * catalog and activate the locale (if it's not already active).
 */
function mergeIntoLinguiCatalog(locale: SupportedLocale, catalog: Catalog) {
  const emptyCatalog: Catalog = { messages: {} };
  const currentCatalog: Catalog = catalogs[locale] || emptyCatalog;
  catalogs[locale] = {
    languageData: catalog.languageData,
    messages: {
      ...currentCatalog.messages,
      ...catalog.messages,
    },
  };
  li18n.load(catalogs);
  li18n.activate(locale);
}
