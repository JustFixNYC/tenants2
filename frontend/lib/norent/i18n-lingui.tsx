import React from "react";
import {
  LoadableCatalog,
  LinguiI18nProps,
  mergeIntoLinguiCatalog,
} from "../i18n-lingui";
import loadable from "@loadable/component";
import i18n, { SupportedLocale } from "../i18n";

const EnCatalog: LoadableCatalog = loadable.lib(
  () => import("../../../locales/en/norent.chunk") as any
);

const EsCatalog: LoadableCatalog = loadable.lib(
  () => import("../../../locales/es/norent.chunk") as any
);

/**
 * Returns a component that loads the Lingui message catalog for
 * the given locale.
 */
function getLinguiCatalogForLanguage(locale: SupportedLocale): LoadableCatalog {
  switch (locale) {
    case "en":
      return EnCatalog;
    case "es":
      return EsCatalog;
  }
}

export const NorentLinguiI18n: React.FC<LinguiI18nProps> = (props) => {
  const locale = i18n.locale;

  const Catalog = getLinguiCatalogForLanguage(locale);

  return (
    <Catalog fallback={<p>Loading locale data...</p>}>
      {(catalog) => {
        mergeIntoLinguiCatalog(locale, catalog);
        return props.children;
      }}
    </Catalog>
  );
};
