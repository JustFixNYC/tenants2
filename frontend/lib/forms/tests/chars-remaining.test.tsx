import React from "react";

import { preloadLingui } from "../../tests/lingui-preloader";
import { LinguiI18n } from "../../i18n-lingui";
import { AppTesterPal } from "../../tests/app-tester-pal";
import { CharsRemaining } from "../chars-remaining";

beforeAll(preloadLingui(LinguiI18n));

describe("<CharsRemaining>", () => {
  it("works when 1 character remains", () => {
    const pal = new AppTesterPal(<CharsRemaining max={15} current={14} />);
    expect(pal.rr.container.firstChild).toMatchSnapshot();
  });

  it("works when many characters remain", () => {
    const pal = new AppTesterPal(<CharsRemaining max={15} current={1} />);
    expect(pal.rr.container.firstChild).toMatchSnapshot();
  });

  it("supports use of spans", () => {
    const pal = new AppTesterPal(
      <CharsRemaining max={15} current={1} useSpan />
    );
    expect(pal.rr.container.firstChild).toMatchSnapshot();
  });
});
