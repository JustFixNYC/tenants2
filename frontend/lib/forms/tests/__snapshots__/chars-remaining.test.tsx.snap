// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CharsRemaining> supports use of spans 1`] = `
<span>
  14 characters remaining
</span>
`;

exports[`<CharsRemaining> works when 1 character remains 1`] = `
<p
  class="has-text-danger"
>
  1 character remaining
</p>
`;

exports[`<CharsRemaining> works when many characters remain 1`] = `
<p>
  14 characters remaining
</p>
`;
